### DISCUSSION TITLE
How to achieve flow distribution with MOOSE framework.

### DISCUSSION THREAD

**Discussion by Raven-pro** [idaholab/moose#23084] [Q&A General] [answered]
Hello, everyone!

I am using THM module and MultiApp to simulate parallel flow channels.

In the parallel flow channels, the pressure drops should be equal in all the channels, which means I need to adjust the mass flow rate of each channel in each time step.

Like, in each time step, the program calculates the pressure drop of each channel and adjusts the inlet mass flow of each channel, then iterates, until the pressure drop of each channel is equal, then we go to the next time step.

So I need to add an iteration at each time step with all the inlet variables and solutions needed, then change the inlets of all the channels to execute the next time step. Are there any possible ways to achieve this?

**Comment by Raven-pro** [comment]
I have also noticed that THM module tutorial doesn't mention something like parallel follow. 

But I think the module is capable of solving such a problem.

Are there any existing components that can achieve this?

**Comment by GiudGiud** [comment]
@joshuahansel @licharlot 

I can see two ways of doing this:
- set an inlet and outlet pressure boundary condition. This will make sure the mass flow rate in each channel does lead to the same pressure drop
- use the optimization module to set up an optimization problem for the mass flow rates to match the pressure drops wanted. This is much more work

**Reply by joshuahansel** [reply]
- That's an interesting idea. I'm not sure what would happen, since the characteristics aren't specified completely at the inlet. It's very easy to try though, so @89018580 you can see what happens.
- This does sound hard.

**Reply by GiudGiud** [reply]
For option 1, can you specify temperature at the inlet for example? Or does it have to be an energy flow. 
I d also be worried about computing density(T) at the inlet without a temperature BC.

**Reply by joshuahansel** [reply]
There's `InletStagnationPressureTemperature1Phase` that might be worth a try (with `Outlet1Phase` at the outlet), but I'm not sure how much the static pressure at the inlet will vary from the stagnation pressure.

**Comment by joshuahansel** [comment, marked as answer]
I need to know more about the problem you're solving. Are the parallel channels connected in any way, i.e., do you have THM junctions connecting them? (Or some other method of "connecting" them?) Or are they all being treated independently?

Do the pressure drops need to be exactly the same for some numerical reason? Or do you just know that they should physically be the same given different paths between the same two end points? The latter reasoning has the requirement that these channels actually be connected somehow. If you have some kind of junction connecting these channels, equal pressure drops result automatically. However, note that this is dependent on the junction formulation, and note that the pressures at the inlet and outlet boundaries may not be in agreement with the corresponding cell-center pressures of the boundary cells. This discrepancy is a function of the spatial discretization error - use slope reconstruction to minimize it.

If you're not using junctions of any kind, one idea is that you can use PID controllers to control the inlet mass flow rate based on a pressure drop target, but you probably don't know what this target is beforehand.

**Reply by joshuahansel** [reply]
But I want to stress that the best solution, in my opinion, is to have some kind of junction connecting the channels that naturally produces the desired condition. Try `JunctionParallelChannels1Phase` for example.

**Reply by licharlot** [reply]
I agree we need to know more. This sounds like a plenum to plenum calculation where the total mass flow rate is known and needs to be split between the parallel channels. In that case, the easiest is indeed to use `JunctionParallelChannels1Phase` but this will require to have all the parallel channels in one input file.
